### PR TITLE
Fixes 15566: Leading literal strings are duplicated when template resolution fails

### LINKED ISSUE #15566: Docker Inspect Issue: Leading literal strings are duplicated when template resolution fails.
When you're using literal strings in a "docker inspect" format string, it seems that if a template resolution fails, leading literal strings can be duplicated. First, an example when it works, both without and with a leading literal string:

sh-4.3# docker inspect -f '{{.Architecture}}' httpd
amd64
sh-4.3# docker inspect -f 'My arch is {{.Architecture}}' httpd
My arch is amd64
sh-4.3#

But look what happens if i misspell as "Archtecture" in both cases:

sh-4.3# docker inspect -f '{{.Archtecture}}' httpd
<no value>
sh-4.3# docker inspect -f 'My arch is {{.Archtecture}}' httpd
My arch is My arch is <no value>
           ^^^^^^^^^^ ???

This is entirely reproducible; just misspell the contents of the template:

sh-4.3# docker inspect -f 'My ports are {{.Config.Ports}}' httpd
My ports are My ports are <no value>

uname -a 
Linux dhcp-25-141.bos.redhat.com 4.0.4-303.fc22.x86_64 #1 SMP Thu May 28 12:37:06 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

docker version
Client:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   650feb2-dirty
 Built:        Thu Aug 13 16:50:09 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   650feb2-dirty
 Built:        Thu Aug 13 16:50:09 UTC 2015
 OS/Arch:      linux/amd64

docker info
Containers: 8
Images: 22
Storage Driver: devicemapper
 Pool Name: docker-8:3-2360227-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.513 GB
 Data Space Total: 107.4 GB
 Data Space Available: 40.34 GB
 Metadata Space Used: 2.228 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93 (2015-01-30)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.0.4-303.fc22.x86_64
Operating System: Fedora 22 (Twenty Two)
CPUs: 4
Total Memory: 11.72 GiB
Name: dhcp-25-141.bos.redhat.com
ID: NODU:OCAV:EOZD:CWD3:EEKS:AJCL:ZCKH:B6S7:7247:UVAC:2EAK:3FC2

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Fixes #15566 
Signed-off-by: Shishir Mahajan shishir.mahajan@redhat.com

### CODE REVIEW

**Comment by icecrime** [comment]
It's pretty obvious from the code that this happens because each template is `Execute`d twice: do you you why this is done in the first place?

**Comment by icecrime** [comment]
This was introduced in #13141, and if I understand the initial intention there, what it means is that we should `Execute` a second time _only_ if the first way of doing things fails (which is not what the code does).

One sure thing: this patch is not the right fix.

Ping @vieux!

**Comment by shishir-a412ed** [comment]
@icecrime 

Yeah All it is doing is, applying the template to the JSON. If there is an error or no value found in the JSON (which it treats as an error), it goes to the condition `err != nil`, where the developer did a `seek` to the start of the JSON again, and applied the template to the entire JSON again.

I assumed that the developer had a good reason (use-case) to do this, so instead of removing that. I segregated the leading string from the template. So that even if it is applied twice, leading string literal would not be part of the template.

Fixing the root cause (why the template is applied twice) would be the right thing. 
I ll be happy to fix this, once we know why it was done in the first place.

Shishir

**Comment by cpuguy83** [comment]
@shishir-a412ed Thanks! I'm going to close this since it is not mergable.

I've also tried to clean this up (#15032) and ran into issues.
The unmarshalling to the raw interface is definitely needed or a larger refactor needs to be done (specifically around runconfig).

**Comment by cpuguy83** [comment]
re-opening and setting to design review to discuss alternatives

**Comment by duglin** [comment]
An easy fix is to just return the error:
```
# git diff | more
diff --git a/api/client/ps/custom.go b/api/client/ps/custom.go
index 261493a..3bff7d8 100644
--- a/api/client/ps/custom.go
+++ b/api/client/ps/custom.go
@@ -200,7 +200,11 @@ func customFormat(ctx Context, containers []types.Container
) {
 		if len(header) == 0 {
 			// if we still don't have a header, we didn't have any containers so we need to fake it to get the right headers from the template
 			containerCtx := &containerContext{}
-			tmpl.Execute(bytes.NewBufferString(""), containerCtx)
+			err := tmpl.Execute(bytes.NewBufferString(""), containerCtx)
+			if err != nil {
+				buffer = bytes.NewBufferString(fmt.Sprintf("Template parsing error: %v\n", err))
+				buffer.WriteTo(ctx.Output)
+			}
 			header = containerCtx.fullHeader()
 		}
```
so we'll now see:
```
# docker ps --format "hello {{ .idx }}"
Template parsing error: template: :1:9: executing "" at <.idx>: idx is not a field of struct type *ps.containerContext
```

**Comment by duglin** [comment]
sorry - was thinking of 'ps' by mistake. will remove comments.

**Comment by duglin** [comment]
I believe this was done because the template exec code is case sensitive. So, when the format uses `{{.Id}}` but the Go struct has it as "ID" then an error is generated.  The code will then redo the template.exec using the "raw" Json which has it as `Id`, and then things work.  And in the case of having some "leading text", it writes that before it sees the error, that's why we see it twice.

I see two options:
1 - decide which is correct: ID or Id and force people to use the correct one. Which may mean change our testcases.  IMO "Id" is the correct one since that's what's defined by our APIs and is in the JSON returned from the daemon.

2 - keep things as they are today (ambiguous and error prone :-) ), but modify the code so that instead of having the template.exec write directly to `cli.out`, have it write to a temp buffer first so that we don't get the double text mentioned in the issue. Then upon a successful template.exec we can write the buffer to cli.out.  And add comment explaining this logic!  And tests!

I'd prefer 1 and clean things up.  It would also make it consistent with 'ps' which is case sensitive.

**Comment by rhatdan** [comment]
I like #1 also.

**Comment by cpuguy83** [comment]
The problem there is that types.Container is embedding runconfig structs even though types.Container should be "safe" to use. Changes to runconfig changes the API response and is not properly versioned.

**Comment by duglin** [comment]
Ideally any changes here should not change the API/JSON on the wire.  If we need to change internal structs then I think that's ok as long as we have the appropriate `'json:"Id"'` type of metadata then we may be able to do it.

**Comment by duglin** [comment]
@shishir-a412ed wanna give it a try?

**Comment by cpuguy83** [comment]
@duglin But `--format` is run against the non-versioned type, not the json wire format. So setting this does not help with `--format`

**Comment by duglin** [comment]
@cpuguy83 I was assuming we'd switch it to use the json data (the 'raw' var in the code)

**Comment by shishir-a412ed** [comment]
@duglin I ll work on it. If I have any questions, I ll ping you on docker-dev.

**Comment by duglin** [comment]
@shishir-a412ed any update on this one?

**Comment by shishir-a412ed** [comment]
@duglin 
Please check now.

File: integration-cli/docker_api_containers_test.go [hunk 1]
```diff
@@ -552,7 +552,7 @@ func (s *DockerSuite) TestContainerApiCommit(c *check.C) {
 	if err != nil {
 		c.Fatal(err)
 	}
-	if cmd != "{[/bin/sh -c touch /test]}" {
```
Comment (duglin) [inline integration-cli/docker_api_containers_test.go]: While I agree with the new format (no {}'s) do you know why they appeared there in the first place?

Comment (shishir-a412ed) [inline integration-cli/docker_api_containers_test.go]: Not completely sure. But my guess is `{}` was to handle the case when the template was applied again in case `err != nil`. Once that code was removed, this `{}` was not needed either.

Comment (duglin) [inline integration-cli/docker_api_containers_test.go]: ok I think I figured out what's going on.

What I think was happening was that `{{Config.Cmd}}` is a `stringutils.StrSlice`, which is defined as
```
type StrSlice struct {
    parts []string
}
```
Since the slice of strings are wrapped by a `struct` we get the extra `{}` around the cmd. Now, we don't see this in the JSON on the wire because we have a specialized JSON serializer func to hide it ( see https://github.com/docker/docker/blob/master/pkg/stringutils/strslice.go#L20 ).

However, I don't think the template.execute() stuff in the CLI code uses that specialized serializer, it just dumps the data as is - meaning with the struct wrapper.  If we do: `docker inspect -f "{{.Config.Cmd.Slice}}" ubuntu` and explicitly step into the struct to the `[]string` via the Slice() func then we'll get the correct results - meaning no `{}`. 

Net: While this PR changes the output of the template.execute() stuff, I think the change is correct. It does highlight a problem though, there could be other cases where using the `-f` option on inspect could produce the wrong output if the internal data type doesn't exactly match what's seen on the wire and a custom (de)serializer is used).  In the long run we should probably stop having the CLI use the same structs as the daemon and just use vanilla data types. We did a little bit of this when we created the api/types/types.go stuff but we may need to go further and do it for the runconfig structs too since those are pulled in.

Anyway, I feel better now that I think I know why this change happened.

Need to check one other change tho.....

Comment (cpuguy83) [inline integration-cli/docker_api_containers_test.go]: We should make sure the internal data structure is not exposed to `-f` in general.

Comment (duglin) [inline integration-cli/docker_api_containers_test.go]: Yup - which means, line like: https://github.com/docker/docker/blob/master/graph/service.go#L46 would need to be replaced with code that converts the "internal" runconfig.Config into something that matches what we want people to see (on the wire and via "inspect -f")

Comment (duglin) [inline integration-cli/docker_api_containers_test.go]: Or, put another way, when we see lines like: https://github.com/docker/docker/blob/master/api/types/types.go#L105  should replace the ref to runconfig.Config with either an in-line struct or another struct defined within that same file (making it clear that its just for exporting/serialization and not internal), and then that would force us to then write the code to map runconfig.Config to this new struct.  In the end, it seems like nothing in types.go should reference a struct/type defined outside of that file.

Comment (cpuguy83) [inline integration-cli/docker_api_containers_test.go]: Unfortunately runconfig is both an internal and an external data structure. (and image config).
It's used as the API types for creating new containers as well as stored on disk (but only after it's mangled a bit).
It's really a nasty mess :(

File: integration-cli/docker_cli_build_test.go [hunk 2]
```diff
@@ -2247,7 +2247,7 @@ func (s *DockerSuite) TestBuildCmd(c *check.C) {
 
 func (s *DockerSuite) TestBuildExpose(c *check.C) {
 	name := "testbuildexpose"
-	expected := "map[2375/tcp:{}]"
+	expected := "map[2375/tcp:map[]]"
```
Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: I believe this is being shown as an empty map rather than as an empty struct because we're only looking at the raw JSON. According to the golang docs on JSON Unmarshal, generic JSON objects (like {}) will be converted to a map - which is what we're seeing above.

Not sure yet which way to go on this. I see two options:
1 - leave it like this and do some hand waving about how Go deals with JSON serialization.
2 - modify the API types along the lines of what was discussed above. Meaning, make the API types match exactly what we want to see on the wire and what the template stuff should produce.   So, @shishir-a412ed when you first mentioned modifying the data structures, that may be right approach, but it would need to only be done for the externalized data structures and not the internal ones.  Then we can deserialize into the real types and use those in the template stuff, and not the raw json.

Unless people think showing 'map' is correct here?  Feels wrong though.

Need to think about this some more... but I'm thinking something more drastic like option 2 might be the right approach.

File: integration-cli/docker_cli_build_unix_test.go [hunk 3]
```diff
@@ -28,8 +28,13 @@ func (s *DockerSuite) TestBuildResourceConstraintsAreUsed(c *check.C) {
...
 	cID := strings.TrimSpace(out)
 
+	cfg, err := inspectFieldJSON(cID, "HostConfig")
+	if err != nil {
+		c.Fatal(err)
+	}
+
 	type hostConfig struct {
-		Memory     int64
+		Memory     float64
```
Comment (duglin) [inline integration-cli/docker_cli_build_unix_test.go]: This could be another example of why we may have to bite the bullet and covert the externalized types. The fact that this shows up as a float instead of an int is really an artifact of Go's JSON deserializing code and I'm not convinced we should expect users to know or care about that.

**Comment by shishir-a412ed** [comment]
Initially just to fix my use-case, I segregated the leading literal strings from the format.

e.g. `-f 'Container Status is {{.State.Status}}'`. I segregated it into two parts:
        a) Container Status is
        b) {{.State.Status}}

So only (b) gets applied to the JSON. With this even if you apply the template twice (to fulfill you usecase), the leading literal strings will never be duplicated. IMHO I never really understood why we need to have the leading literal strings as part of the template?

Now contemplating about your dilemma ((1)whether to fix the external facing structs or (2) just rely on go generic deserializer). I don't think I have a really strong inclination towards one option. I am okay with either one of them. 

With go JSON deserializing code, yes it might use a `float64` instead of an `int64`, or a `map` instead of an `interface{}` but I don't think this concerns the user. But if you feel more comfortable not relying on golang deserializer, and going for option (1), I am okay with that too.

To Summarize:

I am okay with my original implementation (As it's a small patch which fixes the problem and doesn't touch your code). Or with any of the other two options as well.

**Comment by duglin** [comment]
Just testing to see what it might look like I replaced types/ImageInspect's `runconfig.Config` with a local RunConfig struct and it does seem to work to use the types struct instead of the raw json.

Here's the diff (notice I let any possible error from the template exec just show thru):
```
root@docker:~/docker# git diff -w
diff --git a/api/client/inspect.go b/api/client/inspect.go
index 6e728bd..d62bf1a 100644
--- a/api/client/inspect.go
+++ b/api/client/inspect.go
@@ -103,14 +103,19 @@ func (cli *DockerCli) CmdInspect(args ...string) error {
                                        continue
                                }
                                if err := tmpl.Execute(cli.out, inspPtr); err !=
+                                       /*
                                                rdr.Seek(0, 0)
                                                var raw interface{}
                                                if err := dec.Decode(&raw); err 
                                                        return err
                                                }
                                                if err = tmpl.Execute(cli.out, r
+                                       */
                                        return err
+                                       /*
                                                }
+                                       */
+
                                }
                        } else {
                                inspPtr := types.ContainerJSON{}
diff --git a/api/types/types.go b/api/types/types.go
index 66c3f82..c65d7b1 100644
--- a/api/types/types.go
+++ b/api/types/types.go
@@ -91,18 +91,43 @@ type GraphDriverData struct {
        Data map[string]string
 }
 
+type RunConfig struct {
+       Hostname        string
+       Domainname      string
+       User            string
+       AttachStdin     bool
+       AttachStdout    bool
+       AttachStderr    bool
+       ExposedPorts    map[string]struct{} // List of exposed ports
+       PublishService  string
+       Tty             bool
+       OpenStdin       bool
+       StdinOnce       bool
+       Env             []string
+       Cmd             []string
+       Image           string
+       Volumes         map[string]struct{}
+       VolumeDriver    string
+       WorkingDir      string
+       Entrypoint      []string
+       NetworkDisabled bool
+       MacAddress      string
+       OnBuild         []string
+       Labels          map[string]string
+}
+
 // ImageInspect contains response of Remote API:
 // GET "/images/{name:.*}/json"
 type ImageInspect struct {
-       ID              string `json:"Id"`
+       Id              string
        Parent          string
        Comment         string
        Created         string
        Container       string
-       ContainerConfig *runconfig.Config
+       ContainerConfig RunConfig
        DockerVersion   string
        Author          string
-       Config          *runconfig.Config
+       Config          RunConfig
        Architecture    string
        Os              string
        Size            int64
@@ -270,7 +295,7 @@ type ContainerJSONBase struct {
 type ContainerJSON struct {
        *ContainerJSONBase
        Mounts []MountPoint
-       Config *runconfig.Config
+       Config *runconfig.Config // RunConfig
 }
 
 // ContainerJSONPre120 is a backcompatibility struct along with ContainerConfig
@@ -284,7 +309,7 @@ type ContainerJSONPre120 struct {
 
 // ContainerConfig is a backcompatibility struct used in ContainerJSONPre120
 type ContainerConfig struct {
-       *runconfig.Config
+       *runconfig.Config // RunConfig
 
        // backward compatibility, they now live in HostConfig
        Memory     int64
diff --git a/graph/service.go b/graph/service.go
index 1b347d0..fa44c54 100644
--- a/graph/service.go
+++ b/graph/service.go
@@ -35,21 +35,77 @@ func (s *TagStore) Lookup(name string) (*types.ImageInspect,
        }
 
        imageInspect := &types.ImageInspect{
-               ID:              image.ID,
+               Id:        image.ID,
                Parent:    image.Parent,
                Comment:   image.Comment,
                Created:   image.Created.Format(time.RFC3339Nano),
                Container: image.Container,
-               ContainerConfig: &image.ContainerConfig,
+               ContainerConfig: types.RunConfig{
+                       Hostname:     image.ContainerConfig.Hostname,
+                       Domainname:   image.ContainerConfig.Domainname,
+                       User:         image.ContainerConfig.User,
+                       AttachStdin:  image.ContainerConfig.AttachStdin,
+                       AttachStdout: image.ContainerConfig.AttachStdout,
+                       AttachStderr: image.ContainerConfig.AttachStderr,
+                       // ExposedPorts:    image.ContainerConfig.ExposedPorts,
+                       ExposedPorts:    map[string]struct{}{},
+                       PublishService:  image.ContainerConfig.PublishService,
+                       Tty:             image.ContainerConfig.Tty,
+                       OpenStdin:       image.ContainerConfig.OpenStdin,
+                       StdinOnce:       image.ContainerConfig.StdinOnce,
+                       Env:             image.ContainerConfig.Env,
+                       Cmd:             image.ContainerConfig.Cmd.Slice(),
+                       Image:           image.ContainerConfig.Image,
+                       Volumes:         image.ContainerConfig.Volumes,
+                       VolumeDriver:    image.ContainerConfig.VolumeDriver,
+                       WorkingDir:      image.ContainerConfig.WorkingDir,
+                       Entrypoint:      image.ContainerConfig.Entrypoint.Slice(
+                       NetworkDisabled: image.ContainerConfig.NetworkDisabled,
+                       MacAddress:      image.ContainerConfig.MacAddress,
+                       OnBuild:         image.ContainerConfig.OnBuild,
+                       Labels:          image.ContainerConfig.Labels,
+               },
                DockerVersion: image.DockerVersion,
                Author:        image.Author,
-               Config:          image.Config,
+               Config: types.RunConfig{
+                       Hostname:     image.Config.Hostname,
+                       Domainname:   image.Config.Domainname,
+                       User:         image.Config.User,
+                       AttachStdin:  image.Config.AttachStdin,
+                       AttachStdout: image.Config.AttachStdout,
+                       AttachStderr: image.Config.AttachStderr,
+                       // ExposedPorts:    image.Config.ExposedPorts,
+                       ExposedPorts:    map[string]struct{}{},
+                       PublishService:  image.Config.PublishService,
+                       Tty:             image.Config.Tty,
+                       OpenStdin:       image.Config.OpenStdin,
+                       StdinOnce:       image.Config.StdinOnce,
+                       Env:             image.Config.Env,
+                       Cmd:             image.Config.Cmd.Slice(),
+                       Image:           image.Config.Image,
+                       Volumes:         image.Config.Volumes,
+                       VolumeDriver:    image.Config.VolumeDriver,
+                       WorkingDir:      image.Config.WorkingDir,
+                       Entrypoint:      image.Config.Entrypoint.Slice(),
+                       NetworkDisabled: image.Config.NetworkDisabled,
+                       MacAddress:      image.Config.MacAddress,
+                       OnBuild:         image.Config.OnBuild,
+                       Labels:          image.Config.Labels,
+               },
                Architecture: image.Architecture,
                Os:           image.OS,
                Size:         image.Size,
                VirtualSize:  s.graph.GetParentsSize(image) + image.Size,
        }
 
+       for k, v := range image.ContainerConfig.ExposedPorts {
+               imageInspect.ContainerConfig.ExposedPorts[string(k)] = v
+       }
+
+       for k, v := range image.Config.ExposedPorts {
+               imageInspect.Config.ExposedPorts[string(k)] = v
+       }
+
        imageInspect.GraphDriver.Name = s.graph.driver.String()
 
        graphDriverData, err := s.graph.driver.GetMetadata(image.ID)
```

Also, didn't run all of the tests so I don't know for sure what might break but doing:
```
docker inspect -f "{{.Config.Cmd}}" image
```
does work - meaning it shows the CMD w/o {}'s.
And note that `{{.ID}}` won't work, but `{{.Id}}` does, which I think is correct since that's what the JSON shows and we should have the struct align with the json.

Thoughts?  @cpuguy83 @shishir-a412ed

**Comment by shishir-a412ed** [comment]
@duglin PTAL now.

File: api/types/types.go [hunk 4]
```diff
@@ -92,19 +92,44 @@ type GraphDriverData struct {
...
+	Env             []string
+	Cmd             []string
+	Image           string
+	Volumes         map[string]struct{}
+	WorkingDir      string
+	Entrypoint      []string
+	NetworkDisabled bool
+	MacAddress      string
+	OnBuild         []string
+	Labels          map[string]string
```
Comment (duglin) [inline api/types/types.go]: I think you are missing `StopSignal string`

File: daemon/inspect.go [hunk 5]
```diff
@@ -26,7 +26,35 @@ func (daemon *Daemon) ContainerInspect(name string) (*types.ContainerJSON, error
...
+		Env:             container.Config.Env,
+		Cmd:             container.Config.Cmd.Slice(),
+		Image:           container.Config.Image,
+		Volumes:         container.Config.Volumes,
+		WorkingDir:      container.Config.WorkingDir,
+		Entrypoint:      container.Config.Entrypoint.Slice(),
+		NetworkDisabled: container.Config.NetworkDisabled,
+		MacAddress:      container.Config.MacAddress,
+		OnBuild:         container.Config.OnBuild,
+		Labels:          container.Config.Labels,
```
Comment (duglin) [inline daemon/inspect.go]: StopSignal

File: daemon/inspect_unix.go [hunk 6]
```diff
@@ -36,8 +36,36 @@ func (daemon *Daemon) ContainerInspectPre120(name string) (*types.ContainerJSONP
...
+		Env:             container.Config.Env,
+		Cmd:             container.Config.Cmd.Slice(),
+		Image:           container.Config.Image,
+		Volumes:         container.Config.Volumes,
+		WorkingDir:      container.Config.WorkingDir,
+		Entrypoint:      container.Config.Entrypoint.Slice(),
+		NetworkDisabled: container.Config.NetworkDisabled,
+		MacAddress:      container.Config.MacAddress,
+		OnBuild:         container.Config.OnBuild,
+		Labels:          container.Config.Labels,
```
Comment (duglin) [inline daemon/inspect_unix.go]: StopSignal

File: graph/service.go [hunk 7]
```diff
@@ -49,20 +49,72 @@ func (s *TagStore) Lookup(name string) (*types.ImageInspect, error) {
...
+			Env:             image.ContainerConfig.Env,
+			Cmd:             image.ContainerConfig.Cmd.Slice(),
+			Image:           image.ContainerConfig.Image,
+			Volumes:         image.ContainerConfig.Volumes,
+			WorkingDir:      image.ContainerConfig.WorkingDir,
+			Entrypoint:      image.ContainerConfig.Entrypoint.Slice(),
+			NetworkDisabled: image.ContainerConfig.NetworkDisabled,
+			MacAddress:      image.ContainerConfig.MacAddress,
+			OnBuild:         image.ContainerConfig.OnBuild,
+			Labels:          image.ContainerConfig.Labels,
```
Comment (duglin) [inline graph/service.go]: StopSignal

File: graph/service.go [hunk 8]
```diff
@@ -49,20 +49,72 @@ func (s *TagStore) Lookup(name string) (*types.ImageInspect, error) {
...
+			Env:             image.Config.Env,
+			Cmd:             image.Config.Cmd.Slice(),
+			Image:           image.Config.Image,
+			Volumes:         image.Config.Volumes,
+			WorkingDir:      image.Config.WorkingDir,
+			Entrypoint:      image.Config.Entrypoint.Slice(),
+			NetworkDisabled: image.Config.NetworkDisabled,
+			MacAddress:      image.Config.MacAddress,
+			OnBuild:         image.Config.OnBuild,
+			Labels:          image.Config.Labels,
```
Comment (duglin) [inline graph/service.go]: StopSignal

**Comment by duglin** [comment]
@shishir-a412ed I really like the way this reads now - the consistency between the JSON and the struct feels so much better - thanks a lot.

Aside from the nit of the missing StopSignal field (which was just added), my only question is around HostConfig. I noticed that you use the struct version of the properties in the inspect tests, instead of the json format - which is correct. However, I wonder if we should do the same conversion for HostConfig that you did for RunConfig as part of this too - so that we're consistent on that struct as well.

**Comment by duglin** [comment]
ping @cpuguy83 for your take on it too.

**Comment by shishir-a412ed** [comment]
@duglin I have added the `StopSignal` changes.
I don't have any reservations for making the `HostConfig` changes, but I feel it's another big change which would require some more effort. I would rather open a fresh PR for that one and track it there.
WDYT ?

**Comment by duglin** [comment]
@shishir-a412ed a separate PR for HostConfig is fine with me.

just running the tests locally before I LGTM it.

**Comment by duglin** [comment]
@shishir-a412ed TestInspectImage, TestInspectImageFilterInt,TestInspectImageGraphDriver failed for me.

**Comment by shishir-a412ed** [comment]
@duglin last time I tested 3-4 were failing for me. So that seems about right. I ll take a look.

**Comment by shishir-a412ed** [comment]
@duglin check now. All tests should pass.

**Comment by shishir-a412ed** [comment]
PR for fixing the python tests.
https://github.com/docker/docker-py/pull/778

**Comment by shin-** [comment]
What the what? You can't just change the name/capitalization of a JSON field, especially one that's so widely used, especially without ensuring backward compatibility for existing API versions.

:-1:

**Comment by runcom** [comment]
> What the what? You can't just change the name/capitalization of a JSON field, especially one that's so widely used, especially without ensuring backward compatibility for existing API versions.

To solve this, can we version `ImageInspect` like we're doing for `ContainerJSON` at https://github.com/docker/docker/blob/master/api/types/types.go#L280? and add a deprecation notice for `Id` being replaced by `ID`
We can then compare versions and reply with the correct struct (and id field)

**Comment by thaJeztah** [comment]
We now have https://github.com/docker/docker/pull/16564, making it cleaner to version API responses, but agree, only if it's really needed

**Comment by runcom** [comment]
ping @icecrime :)

**Comment by duglin** [comment]
IMO versioning the API is the right approach. I do think we need this since right now we're inconsistent w.r.t. how fields are named between how they appear in the json vs in the inspect `-f` values.

**Comment by calavera** [comment]
This PR has deviated a lot from the original thing it tried to fix. Besides that, this implementation is incorrect, it's modifying versioned types that must remain intact to keep backwards compatibility. I'm closing this, I'm working in a fix for the original bug.
